src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.cpp

```diff
@@ -0,0 +1,52 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+
+/// \cond
+namespace grmhd {
+namespace ValenciaDivClean {
+
+std::array<DataVector, 9> characteristic_speeds(
+    const Scalar<DataVector>& lapse, const tnsr::I<DataVector, 3>& shift,
+    const tnsr::I<DataVector, 3>& spatial_velocity,
+    const Scalar<DataVector>& spatial_velocity_squared,
+    const Scalar<DataVector>& sound_speed_squared,
+    const Scalar<DataVector>& alfven_speed_squared,
+    const tnsr::i<DataVector, 3>& normal) noexcept {
+  auto characteristic_speeds =
+      make_array<9>(DataVector(-1.0 * get(dot_product(normal, shift))));
+
+  const auto& auxiliary_speeds =
```

Comment:
The reference isn't needed here since `characteristic_speeds` is a free function so the RHS is a temporary.